core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -364,30 +368,98 @@ class ReplicaManager(val config: KafkaConfig,
+            val topicPartition = new TopicPartition(topicState.topicName, partitionState.partitionIndex)
+
+            getPartition(topicPartition) match {
+              case HostedPartition.Offline =>
+                stateChangeLogger.warn(s"Ignoring StopReplica request (delete=$deletePartition) from " +
+                  s"controller $controllerId with correlation id $correlationId " +
+                  s"epoch $controllerEpoch for partition $topicPartition as the local replica for the " +
+                  "partition is in an offline log directory")
+                responseMap.put(topicPartition, Errors.KAFKA_STORAGE_ERROR)
+
+              case HostedPartition.Online(partition) =>
+                val currentLeaderEpoch = partition.getLeaderEpoch
+                val requestLeaderEpoch = partitionState.leaderEpoch
+                // When a topic is deleted, the leader epoch is not incremented. To circumvent this,
+                // a sentinel value (EpochDuringDelete) overwriting any previous epoch is used.
+                // When an older version of the StopReplica request which does not contain the leader
+                // epoch, a sentinel value (NoEpoch) is used and bypass the epoch validation.
+                if (requestLeaderEpoch == LeaderAndIsr.EpochDuringDelete ||
+                    requestLeaderEpoch == LeaderAndIsr.NoEpoch ||
+                    requestLeaderEpoch > currentLeaderEpoch) {
+                  partitions += topicPartition
+                } else if (requestLeaderEpoch < currentLeaderEpoch) {
+                  stateChangeLogger.warn(s"Ignoring StopReplica request (delete=$deletePartition) from " +
+                    s"controller $controllerId with correlation id $correlationId " +
+                    s"epoch $controllerEpoch for partition $topicPartition since its associated " +
+                    s"leader epoch $requestLeaderEpoch is smaller than the current " +
+                    s"leader epoch $currentLeaderEpoch")
+                  responseMap.put(topicPartition, Errors.STALE_CONTROLLER_EPOCH)
+                } else {
+                  stateChangeLogger.debug(s"Ignoring StopReplica request (delete=$deletePartition) from " +
+                    s"controller $controllerId with correlation id $correlationId " +
+                    s"epoch $controllerEpoch for partition $topicPartition since its associated " +
+                    s"leader epoch $requestLeaderEpoch matches the current leader epoch")
+                  responseMap.put(topicPartition, Errors.STALE_CONTROLLER_EPOCH)
+                }
+
+              case HostedPartition.None =>
+                try {
+                  // Delete log and corresponding folders in case replica manager doesn't hold them anymore.
+                  // This could happen when topic is being deleted while broker is down and recovers.
+                  maybeCleanReplica(topicPartition, deletePartition)
```

Comment:
This changes the order of the operations. Previously we would have stopped fetchers before attempting to delete the log directory. Are we sure this is safe?

Reply:
I need to take another look at it.

Reply:
I had a second look at this one and I think that the change is safe. Let me explain.

We can end up in this situation where the replica is not know any more by broker in two ways:

1. The broker receives a StopReplica request without having received a LeaderAndIsr request prior to it. In this case, the partition is not created and the fetchers haven't been started so we don't need to stop them for the concerned partitions. We have one test which simulate such scenario: https://github.com/apache/kafka/blob/trunk/core/src/test/scala/unit/kafka/admin/DeleteTopicTest.scala#L432. It fails if I comment the cleaning logic.

2. The handling of the StopReplica request fails with a storage exception when it deletes the log directory. The delete happens after the partition is effectively removed from the `allPartitions` map in the ReplicaManager. Note that the fetchers for the concerned partitions are already stopped at this point as they are stopped before removing the partition from the Map. If the request is retried somehow, the partition won't be there so the cleaning would take place.

All together, fetchers are always started after the partition is added in the `allPartitions` Map and always stopped before removing the partition from the Map. If it is not in the Map, fetchers can't be started. Thus, this seems safe to me based on my current knowledge.

The only benefit of putting it there is that the logging is better in my opinion. When the replica does not exist, we don't get the `handling StopReplica...` and `completed StopReplica...` but only `Ignoring StopReplica... cause replica does not exist`.

I would be fine reverting back to the prior behavior as it is only a cosmetic change at the end. It may be safer to do so.